File: backend/src/main/java/com/acme/salary/common/error/ApiError.java
package com.acme.salary.common.error;

import java.time.Instant;
import java.util.List;
import org.springframework.http.HttpStatus;

/**
 * The single error envelope every failing request returns, as published in the README.
 *
 * <p>It deliberately carries no stack trace, SQL or internal class name (NFR-2.6).
 */
public record ApiError(
        Instant timestamp,
        int status,
        String error,
        String message,
        String path,
        List<FieldError> fieldErrors) {

    /** One per rejected input field, so a client can attach messages to form controls. */
    public record FieldError(String field, String message) {
    }

    public static ApiError of(HttpStatus status, String message, String path) {
        return new ApiError(Instant.now(), status.value(), status.getReasonPhrase(), message, path, List.of());
    }

    public static ApiError of(HttpStatus status, String message, String path, List<FieldError> fieldErrors) {
        return new ApiError(Instant.now(), status.value(), status.getReasonPhrase(), message, path, fieldErrors);
    }
}

File: frontend/src/app/core/http/api-error.ts
/**
 * The API's single error envelope, and the shape this application turns it into.
 *
 * Every failing request returns the same body, so there is exactly one thing to parse —
 * see `com.acme.salary.common.error.ApiError`.
 */

/** One rejected input field, so a message can be attached to the right form control. */
export interface ApiFieldError {
  field: string;
  message: string;
}

/** The envelope as it arrives on the wire. */
export interface ApiErrorBody {
  timestamp: string;
  status: number;
  error: string;
  message: string;
  path: string;
  /** Absent rather than empty when there are none — Jackson omits nulls. */
  fieldErrors?: ApiFieldError[];
}

/**
 * A failed request, normalised.
 *
 * The error interceptor converts every failure into one of these, so a component never
 * has to inspect an `HttpErrorResponse`, guess whether a body is present, or decide what
 * a network-level failure should say. Components read `message` and `fieldErrors`.
 */
export class ApiFailure extends Error {
  constructor(
    /** HTTP status, or 0 when the request never reached the server. */
    readonly status: number,
    message: string,
    readonly fieldErrors: ApiFieldError[] = [],
  ) {
    super(message);
    this.name = 'ApiFailure';
  }

  /** Whether this failure can be explained against specific form controls. */
  get hasFieldErrors(): boolean {
    return this.fieldErrors.length > 0;
  }

  /** The message for one control, or undefined if the failure was not about it. */
  messageFor(field: string): string | undefined {
    return this.fieldErrors.find((candidate) => candidate.field === field)?.message;
  }
}
